### PR TITLE
[XPU][OptEW] Define `-intel-triton-optimize-elementwise-parallelism` pass

### PR DESCRIPTION
**Pull request by victor-eds**
Define pass improving elementwise parallelism by avoiding layout conversions leading to data duplication between threads.

See pass documentation for more information.

### CODE REVIEW

**Comment by victor-eds** [comment]
First step for #2562.

File: test/TritonIntelGPU/optimize-elementwise.mlir [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  // CHECK:           %[[VAL_3:.*]] = arith.addf %[[VAL_0]], %[[VAL_1]] : tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #[[$ATTR_4]]}>>
+  %0 = triton_gpu.convert_layout %arg0 : tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #blocked}>> -> tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #blocked1}>>
+  %1 = triton_gpu.convert_layout %arg1 : tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #blocked}>> -> tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #blocked1}>>
+  // CHECK:           %[[VAL_4:.*]] = tt.splat %[[VAL_2]] : f32 -> tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #[[$ATTR_4]]}>>
+  %2 = tt.splat %arg2 : f32 -> tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #blocked1}>>
+  %3 = arith.addf %0, %1 : tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #blocked1}>>
+  // CHECK:           %[[VAL_5:.*]] = arith.addf %[[VAL_4]], %[[VAL_3]] : tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #[[$ATTR_4]]}>>
```
Comment (victor-eds) [inline test/TritonIntelGPU/optimize-elementwise.mlir:62]: The layout conversion that would result from `%[[VAL_3]]` conversion is erased by the pass as we can see

File: third_party/intel/include/Dialect/TritonIntelGPU/Transforms/Passes.td [hunk 2]
```diff
@@ -365,4 +365,46 @@ tt.func @test(%arg0: tensor<32x32xf32, #mma>) -> tensor<32xf32, #triton_gpu.slic
...
+    %0 = arith.addf %arg0, %arg1 : tensor<16xf32, #triton_gpu.slice<{dim = 1, parent = #blocked}>>
+    %1 = triton_gpu.convert_layout %0 : tensor<16xf32, #triton_gpu.slice<{dim = 1, parent = #blocked}>> -> tensor<16xf32, #triton_gpu.slice<{dim = 1, parent = #blocked1}>>
+    tt.return %1 : tensor<16xf32, #triton_gpu.slice<{dim = 1, parent = #blocked1}>>
+  }
+}
+```
+  }];
+
+  let dependentDialects = ["mlir::triton::TritonDialect",
+                           "mlir::triton::gpu::TritonGPUDialect"];
```
Comment (victor-eds) [inline third_party/intel/include/Dialect/TritonIntelGPU/Transforms/Passes.td:407]: If we find an elementwise operation from dialect `X`, that means the dialect has been loaded already, so no need to have it as a dependent dialect in order to create this. We want to include these two for future proofing, tho.

File: third_party/intel/include/Dialect/TritonIntelGPU/Transforms/Passes.td [hunk 3]
```diff
@@ -365,4 +365,46 @@ tt.func @test(%arg0: tensor<32x32xf32, #mma>) -> tensor<32xf32, #triton_gpu.slic
...
 
+def TritonIntelGPUOptimizeElementwiseParallelism
+    : Pass<"tritonintelgpu-optimize-elementwise-parallelism", "mlir::ModuleOp"> {
+  let summary =
+      "Improve parallelism of elementwise operations utilizing more hardware resources";
+
+  let description = [{
+    Detect elementwise operations with an encoding causing sub-par parallelism,
+    i.e., with data duplication across threads, and convert the operands to a
+    more optimal encoding if the cost of doing so is not too high according to
```
Comment (etiotto) [inline third_party/intel/include/Dialect/TritonIntelGPU/Transforms/Passes.td:376]: "is not too high according to some heuristics" -> "is heuristically estimated to be sufficiently low"

File: test/TritonIntelGPU/optimize-elementwise.mlir [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  // CHECK:           tt.return %[[VAL_3]] : tensor<16xf32, #triton_gpu.slice<{dim = 1, parent = #[[$ATTR_1]]}>>
+  tt.return %2 : tensor<16xf32, #triton_gpu.slice<{dim = 1, parent = #blocked1}>>
+}
+
+// -----
+
+#blocked = #triton_gpu.blocked<{sizePerThread = [1, 16], threadsPerWarp = [16, 1], warpsPerCTA = [1, 1], order = [0, 1]}>
+#blocked1 = #triton_gpu.blocked<{sizePerThread = [16, 1], threadsPerWarp = [1, 16], warpsPerCTA = [1, 1], order = [0, 1]}>
+
+// CHECK: #[[$ATTR_2:.+]] = #triton_gpu.blocked<{sizePerThread = [1, 16], threadsPerWarp = [16, 1], warpsPerCTA = [1, 1], order = [0, 1]}>
```
Comment (etiotto) [inline test/TritonIntelGPU/optimize-elementwise.mlir:27]: [nit]: Call it ATTR_0, ATTR_1 like in the first test.  Same for other tests.

File: test/TritonIntelGPU/optimize-elementwise.mlir [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+// CHECK-SAME:                                       %[[VAL_0:.*]]: tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #[[$ATTR_0]]}>>,
+// CHECK-SAME:                                       %[[VAL_1:.*]]: f32
+tt.func @test_convert_layout_splat(%arg0: tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #blocked}>>, %arg1: f32) -> tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #blocked1}>> {
+  %0 = triton_gpu.convert_layout %arg0 : tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #blocked}>> -> tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #blocked1}>>
+  // CHECK:           %[[VAL_2:.*]] = tt.splat %[[VAL_1]] : f32 -> tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #[[$ATTR_2]]}>>
+  %1 = tt.splat %arg1 : f32 -> tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #blocked1}>>
+  // CHECK:           %[[VAL_3:.*]] = arith.addf %[[VAL_0]], %[[VAL_2]] : tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #[[$ATTR_2]]}>>
+  // CHECK:           %[[VAL_4:.*]] = triton_gpu.convert_layout %[[VAL_3]] : tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #[[$ATTR_2]]}>> -> tensor<128xf32, #triton_gpu.slice<{dim = 1, parent = #[[$ATTR_3]]}>>
```
Comment (etiotto) [inline test/TritonIntelGPU/optimize-elementwise.mlir:38]: The original code and the transformed code both contain 1 convert layout operation. What makes the transformed code "cheaper" ?

**Comment by victor-eds** [comment]
I may change the approach to this next week. Delay reviews for now.

**Review by etiotto — APPROVED** [review]

**Review by chengjunlu — APPROVED** [review]

**Pull request merged**